### DISCUSSION TITLE
Shared linking errors

### DISCUSSION THREAD

**Discussion by drfiemost** [msys2/MINGW-packages#8056] [Q&A] [answered]
I'm trying to make a package for libopenmpt, a mod player library evolved from libmodplug and actively maintained, but facing the following error which prevents the shared lib to be built:

```
*** Warning: linker path does not have real file for library -lz.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libz and none of the candidates passed a file format test
*** using a file magic. Last file checked: C:/msys64/mingw64/lib/libz.dll.a

*** Warning: linker path does not have real file for library -lrpcrt4.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with librpcrt4 and none of the candidates passed a file format test
*** using a file magic. Last file checked: C:/msys64/mingw64/x86_64-w64-mingw32/lib/librpcrt4.a
*** The inter-library dependencies that have been dropped here will be
*** automatically added whenever a program is linked with this library
*** or is declared to -dlopen it.

*** Since this library must not contain undefined symbols,
*** because either the platform does not support them or
*** it was explicitly requested with -no-undefined,
*** libtool will only create a static version of it.
```

Both libraries are installed and recognized like the other:

```
$ file /mingw64/lib/libz.dll.a
/mingw64/lib/libz.dll.a: current ar archive

$ file /mingw64/x86_64-w64-mingw32/lib/librpcrt4.a
/mingw64/x86_64-w64-mingw32/lib/librpcrt4.a: current ar archive
```

What's going wrong? I've tried also running autoreconf -fi in advance but no luck.
 I'm attaching the pkgbuild for reference.

[PKGBUILD.txt](https://github.com/msys2/MINGW-packages/files/6075405/PKGBUILD.txt)

**Comment by Biswa96** [comment, marked as answer]
I can not reproduce your issue in my PC. Maybe something was corrupted in your MSYS2 installer. Try with a new msys2 install or commit the PKGBUILD in your fork repository and let GitHub Actions to build it for you.

**Reply by drfiemost** [reply]
Ok thanks, I will try. Weird thing is that other packages using zlib builds fine.

**Reply by revelator** [reply]
there has been reports of this behaviour on win10 a workaround is to put this in before configuring in PKGBUILD
export lt_cv_deplibs_check_method=${lt_cv_deplibs_check_method='pass_all'}

**Reply by drfiemost** [reply]
Yep, that did the trick! Thanks

**Reply by revelator** [reply]
np we are here to help :) though it is kinda strange that it only seems to affect win10 users.

**Comment by GitMensch** [comment]
> there has been reports of this behaviour on win10 a workaround is to put this in before configuring in PKGBUILD
`export lt_cv_deplibs_check_method=${lt_cv_deplibs_check_method='pass_all'}`

Just to recheck: if this is a common issue (I've just suggested the same to someone with a similar problem that a package on mingw64-shell did not compile any more after passing `--enable-code-coverage` to the configure script - should the export be part of the MSYS2 default profile? It seems a bunch of packages have added this in the PKGBUILD, which seems to be bad: if that is "commonly needed", why not exporting it by default?

**Reply by revelator** [reply]
not a bug as such its a libtool quirk on windows.
libtool uses file.exe on unixes to determine the type eg. dynamic / static arch and a whole slew of other stuff, sadly file.exe has problems detecting the correct types on windows. i use a small bash script to override the file.exe in /usr/bin basically just telling libtool to do the same as the small fix above and ignore the type check completely. we could just ship a package with this script if demand is high enough ?.

**Reply by GitMensch** [reply]
That sounds it would be much better to fix file.exe, no? Do you have an outline somewhere about the problems with that?

**Reply by revelator** [reply]
some of it seems to stem from it sometimes being unable to determine the archive type of libraries -> eg warns about not being able to link to a static library when that library is actually an import library -> somelib.dll.a i think this might be caused by the difference on linux where libtool links to the .so on shared builds and not the import library but im not 100% on that ?. in that case we just skip the type check to allow linking it anyway as a workaround. tbh im not certain if all this mucking around with type checks just makes things more complicated than nessesary ?. sadly we now only build the libltdl part of libtool for mingw-w64 so we cant easily override this behaviour anymore and even worse the mingw-w64 version of file outputs windows paths which libtool does not know how to deal with.

**Reply by revelator** [reply]
the 32 bit bash script i use is here ->

```
#!/bin/sh

###############################################################################
#
# This source is public domain.
#
# Minimalistic 'file' command for libtool on mingw
# This matches libtool's win32_libid shell function in the libtool file itself.
#
###############################################################################

while $(echo "$1" | grep --silent '^-')
do
	shift
done

case "$1" in
	*.exe) echo "MS Windows PE32 Intel 80386 console executable not relocatable" ;; # note: libtool doesn't match the "MS Windows PE 32-bit Intel spec".
	*.dll) echo "MS Windows PE32 Intel 80386 console DLL" ;; # note: libtool doesn't match the "MS Windows PE 32-bit Intel spec".
	*.dll.a) echo "ar archive import library" ;; # <-- import library for relocatable library.
	*.a | *.lib) echo "ar archive" ;; # <-- could be static library or import library for relocatable library; libtool will use objdump to find out (not fail safe).
	*) echo "pass_all" ;;
esac
```

and the 64 bit version

```
#!/bin/sh

###############################################################################
#
# This source is public domain.
#
# Minimalistic 'file' command for libtool on mingw
# This matches libtool's win32_libid shell function in the libtool file itself.
#
###############################################################################

while $(echo "$1" | grep --silent '^-')
do
	shift
done

case "$1" in
	*.exe) echo "MS Windows PE32+ Intel x86-64 console executable not relocatable" ;; # note: libtool doesn't match the "MS Windows PE 64-bit Intel spec".
	*.dll) echo "MS Windows PE32+ Intel x86-64 console DLL" ;; # note: libtool doesn't match the "MS Windows PE 64-bit Intel spec".
	*.dll.a) echo "ar archive import library" ;; # <-- import library for relocatable library.
	*.a | *.lib) echo "ar archive" ;; # <-- could be static library or import library for relocatable library; libtool will use objdump to find out (not fail safe).
	*) echo "pass_all" ;;
esac
```

what they do is skip typechecks for windows imports eg. pass_all on those types everything else is still handled by file.
the scripts are not by me originally, the original was taken from the old msys.org and updated to handle 64 bit imports also.

**Reply by GitMensch** [reply]
Couldn't we use file.exe for .dll and .exe as-is, or use it to check for 32/64bit at least? This would drop the need to have a specific one for 32/64bit and would also show issues like mismatches (the scripts above will always say 32/64bit)

Here is is some of msys2 `/usr/bin/file`'s output:

binaries created by subsystem mingw64
~~~
libpdcurses.a:        current ar archive
libpdcurses.dll.a:    current ar archive
wingui-shared-MINGW64/firework.exe: PE32+ executable (GUI) x86-64 (stripped to external PDB), for MS Windows, 11 sections
libpdcurses_wingui.dll: PE32+ executable (DLL) (GUI) x86-64 (stripped to external PDB), for MS Windows, 11 sections
libpdcurses_wincon.dll: PE32+ executable (DLL) (console x86-64 (stripped to external PDB), for MS Windows, 11 sections
~~~

binaries created by subsystem mingw32
~~~
libpdcurses.a:        current ar archive
libpdcurses.dll.a:    current ar archive
cobc.exe: PE32 executable (console) Intel 80386 (stripped to external PDB), for MS Windows, 9 sections
libcob-4.dll: PE32 executable (DLL) (console) Intel 80386 (stripped to external PDB), for MS Windows, 10 sections
~~~

binaries created by subsystem clang-aarch64:

~~~
libpdcurses_wingui.dll: PE32+ executable (DLL) (GUI) Aarch64, for MS Windows, 8 sections
~~~

binaries created by subsystem ucrt-x86_64:

~~~
libpdcurses_vt.dll: PE32+ executable (DLL) (console) x86-64 (stripped to external PDB), for MS Windows, 11 sections
~~~

**Reply by revelator** [reply]
that would indeed be best but notice this ->
libpdcurses.a:        current ar archive -> this one is detected ok
libpdcurses.dll.a:    current ar archive -> this one is wrong it should be import library...

see what i mean ? :)

**Reply by GitMensch** [reply]
If we find that this is the only issue then we can handle `*.dll.a)` in the wrapper and send the other 3 to file.exe.

**Reply by revelator** [reply]
it should be possible to make some code for file.exe to determine the correct archive type in import archives by checking for __imp_*symbol in the library (these should not exist in the static ones), it wont be foolprof but better than what is i reckon.

**Reply by revelator** [reply]
also a possibility :)

**Reply by revelator** [reply]
one thing to look out for with that approach is that sometimes we see import libraries with the *.a extension instead of *.dll.a (yeah rather annoying i know) the mingw-w64-api for instance has import libraries marked as ar archives but they are actually import libraries for the windows system dll's. not sure why file accepts those but it might be because the dll's are avaliable system wide so it can read them. i have the nasty feeling that file would actually work as intended if the dll's in /where ever/bin are on PATH. i think this needs some investigating.

**Reply by GitMensch** [reply]
fixing file.exe would be preferable to a wrapper (now tracked under https://github.com/msys2/MSYS2-packages/issues/3927), and if

> mingw-w64 version of file outputs windows paths which libtool does not know how to deal with

is true then we can always call the msys one using its fullpath  `/usr/bin/file.exe` in the wrapper that would use the fixed file.exe.

So ... should we come up with a wrapper that uses this approach and for now overrides the result for "dll.a" to say "import library"?

**Reply by GitMensch** [reply]
> i have the nasty feeling that file would actually work as intended if the dll's in /where ever/bin are on PATH. i think this needs some investigating.

that would indeed be interesting, but as this is commonly the case for pacman installed libraries then this should "normally" work, depending how it translates the name to the actual lib (which should be included in the import library, if I don't remember that wrong)

**Reply by revelator** [reply]
ah no it still lists them as ar archive bummer...
the nasty truth is that windows itself does not have anything similar to file it does content checks per extension not by the data in the archive :S msvc for instance has no notion of somelib_shared.lib being any different from somelib_static.lib the developer needs to make sure he uses the correct library there are no safegaurds really. You can ofc. link directly to a dll even in msvc but most of us have gotten used to doing it the hard way and just being carefull.

**Reply by GitMensch** [reply]
> You can ofc. link directly to a dll even in msvc but most of us have gotten used to doing it the hard way and just being carefull.

Are you sure? I've thought that's a GCC only feature.

**Reply by revelator** [reply]
guess ill have to rip open the code for file and having a look at what goes on behind our backs.
i seem to remember it uses objdump for parsing the content of the libraries hmm...

**Reply by revelator** [reply]
perfectly sure :) i done it numerous times.

**Reply by revelator** [reply]
you need a small piece of code invoking loadlibrary on the dll with msvc but perfectly doable, gcc is easier though ;) the dll also needs to be located besides the program using it or on path but the same is true with mingw unless you specifically write code to load it from another location.

**Reply by revelator** [reply]
holy batman file's source code is quite a beast :S still trying to find where in the source it determines library types so far not much luck.
it seems it is using lookups based on bit types in the archive (guess that is what the magic part is for).

**Reply by GitMensch** [reply]
So the issue may be that we don't set the right bits in there to begin with?

**Reply by revelator** [reply]
that may very well be the case or file might hmm. 
not sure where to go from here without getting a hold on the maintainer, he/she might know more about the internals.
